src/python/pants/backend/python/async_utils/sync_to_async.py

```diff
@@ -0,0 +1,20 @@
+"""
+Generic utility that enables asynchronously executing an arbitrary
+synchronous function. Adapted from https://github.com/kumaraditya303/aioshutil.
+"""
+
+import asyncio
+import functools
+from collections.abc import Callable, Coroutine
+from typing import Any
+
+# n.b. Until support for pre-Python 3.10 versions is dropped, ParamSpecs 
+# aren't supported to improve the typing here: https://peps.python.org/pep-0612/
+def wrap(func: Callable) -> Callable[..., Coroutine]:
+  @functools.wraps(func)
+  async def run_async(*args: Any, **kwargs: Any) -> Any:
+    event_loop = asyncio.get_event_loop()
+    runnable = functools.partial(func, *args, **kwargs)
+    return await event_loop.run_in_executor(None, runnable)
+  
+  return run_async
```

Comment:
Ideally we'd write this as:

```python
def wrap(func: Callable[P, R]) -> Callable[P, Coroutine[Any, Any, R]]:
  @functools.wraps(func)
  async def run_async(*args: P.args, **kwargs: P.kwargs) -> R:
    ...
  return run_async
```

but it seems like we're using py3.9 to build pants itself

Reply:
To clarify an important point: Pants has its own async event loop, in Rust. `await Get()` (or, under call-by-name, `await foo()` actually yields execution to that loop. Adding an ad-hoc python event loop to that is not helpful, and may be harmful.

I think we can get rid of all of this and block on a synchronous call to `copytree`.

If it is an issue, the fix is to add an "intrinsic" (i.e., a Rust function with Python bindings) that does a copytree natively.

Reply:
Ah, I didn't realize that - I'm happy to unwind the async utils here and then just make the sync `copytree` call. Thanks for clarifying!